pkg/flexvolume/flexvolume.go

```diff
@@ -0,0 +1,308 @@
+		return fmt.Errorf("jsonOptions is not set by user properly: %#v", jsonOptions)
+	}
+
+	d.cinderConfig = volOptions[knownflags.CinderConfigKey].(string)
+	d.volId = volOptions[knownflags.VolIdKey].(string)
+	// this is a system option
+	d.fsType = volOptions["kubernetes.io/fsType"].(string)
+
+	manager, err := NewFlexManager(d.cinderConfig)
+	if err != nil {
+		return err
+	}
+	d.manager = manager
+
+	return nil
+}
+
+// initFlexVolumeDriverForUnMount use targetMountDir to initialize FlexVolumeDriver from magic file
+func (d *FlexVolumeDriver) initFlexVolumeDriverForUnMount(targetMountDir string) error {
+	// use the magic file to store volId since flexvolume will execute fresh new binary every time
+	optsData, err := utiljson.ReadJsonOptsFile(targetMountDir)
+	if err != nil {
+		return err
+	}
+
+	d.cinderConfig = optsData[knownflags.CinderConfigKey].(string)
+	d.volId = optsData[knownflags.VolIdKey].(string)
+
+	manager, err := NewFlexManager(d.cinderConfig)
+	if err != nil {
+		return err
+	}
+	d.manager = manager
+
+	return nil
+}
+
+// Invocation: <driver executable> attach <json options> <node name>
+func (d *FlexVolumeDriver) attach(jsonOptions, nodeName string) (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// Invocation: <driver executable> detach <mount device> <node name>
+func (d *FlexVolumeDriver) detach(mountDev, nodeName string) (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// Invocation: <driver executable> waitforattach <mount device> <json options>
+func (d *FlexVolumeDriver) waitForAttach(mountDev, jsonOptions string) (map[string]interface{}, error) {
+	return map[string]interface{}{"device": mountDev}, nil
+}
+
+// Invocation: <driver executable> isattached <json options> <node name>
+func (d *FlexVolumeDriver) isAttached(jsonOptions, nodeName string) (map[string]interface{}, error) {
+	return map[string]interface{}{"attached": true}, nil
+}
+
+// Invocation: <driver executable> mount <mount dir> <json options>
+// mount will:
+// 1. attach Cinder volume to target dir by AttachDisk
+// 2. store meta data generated by AttachDisk into a json file in target dir
+func (d *FlexVolumeDriver) mount(targetMountDir, jsonOptions string) (map[string]interface{}, error) {
+	glog.V(5).Infof("Cinder flexvolume mount %s to %s", d.volId, targetMountDir)
+
+	// initialize cinder driver from user provided jsonOptions
+	if err := d.initFlexVolumeDriverForMount(jsonOptions); err != nil {
+		return nil, err
+	}
+
+	// attach cinder disk to host machine
+	if err := d.manager.AttachDisk(d, targetMountDir); err != nil {
+		glog.V(4).Infof("AttachDisk failed: %v", err)
+		return nil, err
+	}
+	glog.V(3).Infof("Cinder volume %s attached", d.volId)
+
+	// NOTE(harry) mount a tmpfs in targetMountDir. This is because when doing unmount,
+	// k8s will check if this dir is a mount point, if not, it will bypass the umount call
+	// This has been fixed in #49118, then we can delete this workaround in new release.
```

Comment:
since we already have a workaround of #49118, do we still need to set `--enable-controller-attach-detach=false`?